server/src/com/cloud/deploy/DeploymentPlanningManagerImpl.java

```diff
@@ -584,105 +583,64 @@ private void checkForNonDedicatedResources(VirtualMachineProfile vmProfile, Data
-                    allHostsFromDedicatedID.add(vo.getHostId());
-                }
+            //Only when the type is instance VM and not explicitly dedicated.
+            if (vm.getType() == VirtualMachine.Type.User && !isExplicit) {
+                //add explicitly dedicated resources in avoidList
 
-                //Remove the dedicated ones from main list
-                allPodsInDc.removeAll(allPodsFromDedicatedID);
-                allClustersInDc.removeAll(allClustersFromDedicatedID);
-                allHostsInDc.removeAll(allHostsFromDedicatedID);
+                avoids.addPodList(allPodsInDc);
+                avoids.addClusterList(allClustersInDc);
+                avoids.addHostList(allHostsInDc);
             }
-            else {
-                //The dedicated resource belongs to VM Domain ID or No dedication.
 
-                tempStorage = _dedicatedDao.searchDedicatedPods(null, vmDomainId, null, null).first();
+            //Handle the Virtual Router Case
+            //No need to check the isExplicit. As both the cases are handled.
+            if (vm.getType() == VirtualMachine.Type.DomainRouter) {
+                long vmAccountId = vm.getAccountId();
+                long vmDomainId = vm.getDomainId();
 
-                for(DedicatedResourceVO vo : tempStorage) {
-                    allPodsFromDedicatedID.add(vo.getPodId());
-                }
+                List<Long> allPodsFromDedicatedID = new ArrayList<Long>();
+                List<Long> allClustersFromDedicatedID = new ArrayList<Long>();
+                List<Long> allHostsFromDedicatedID = new ArrayList<Long>();
 
-                tempStorage.clear();
-                tempStorage = _dedicatedDao.searchDedicatedClusters(null, vmDomainId, null, null).first();
-
-                for(DedicatedResourceVO vo : tempStorage) {
-                    allClustersFromDedicatedID.add(vo.getClusterId());
-                }
+                for (DedicatedResourceVO vo : _dedicatedDao.listAvailableResources(vmAccountId, vmDomainId)) {
+                    if (vo.getPodId() != null) {
+                        allPodsFromDedicatedID.add(vo.getPodId());
+                        continue;
+                    }
 
-                tempStorage.clear();
-                tempStorage = _dedicatedDao.searchDedicatedHosts(null, vmDomainId, null, null).first();
+                    if (vo.getClusterId() != null) {
+                        allClustersFromDedicatedID.add(vo.getClusterId());
+                        continue;
+                    }
 
-                for(DedicatedResourceVO vo : tempStorage) {
-                    allHostsFromDedicatedID.add(vo.getHostId());
+                    if (vo.getHostId() != null) {
+                        allHostsFromDedicatedID.add(vo.getHostId());
+                    }
                 }
 
                 //Remove the dedicated ones from main list
                 allPodsInDc.removeAll(allPodsFromDedicatedID);
```

Comment:
Can you put some explanation as to why there is a difference in checking for dedicated resources for user VM and VR? In case of user VMs dedicated resources are put in avoid list, for VR only non-dedicated resources are put in avoid list.